### ISSUE TITLE
member list doesn't respond to non-full cluster ("context deadline exceeded")

### ISSUE THREAD

**Issue by eaoliver** [coreos/etcd#2849]
I specified a five-node cluster using the discovery service.  Four nodes are started and visible to each other, as verified in the journal.  However, no node can view the existing member list.

core@vb-command ~ $ etcdctl --debug member list
context deadline exceeded
core@vb-command ~ $ etcdctl --version
etcdctl version 2.0.10

On all nodes: etcdctl ls blocks until all five nodes are in the cluster.  Is this documented anywhere?  This design is not very flexible.  A 4-active-node cluster should be fully functional, even with a non-existing fifth (presumably failed) node.

**Comment by xiang90** [comment]
@eaoliver This is a bug, not by design. I have found the cause, it should be fixed by next week with our new release.

**Comment by xiang90** [comment]
@eaoliver Another thing to mention: you need to give your etcdctl a full list of your members. Or it will only look for the localhost:4001. And if that one fails, it is not able to find anything else.

**Comment by eaoliver** [comment]
@xiang90 I'm using the discovery service and all nodes are running etcd2 and can access the Internet.  I don't manipulate the membership list unless I am removing a node.

**Comment by xiang90** [comment]
@eaoliver I mean for etcdctl. Or how will etcdctl knows where is your cluster? You need to configure it properly.

**Comment by eaoliver** [comment]
@xiang90 This is a basic deployment and all nodes are running CoreOS alpha (681.0.0) and have access to the discovery service. I've followed all the documentation that is available for etcd2.  What am I missing in the cloud config?

coreos:
   etcd2:
      discovery: https://discovery.etcd.io/86fed56b755e07e0f4d1aa7e367f0841
      listen-client-urls: http://localhost:2379
      advertise-client-urls: http://localhost:2379
      listen-peer-urls: http://$private_ipv4:2380
      initial-advertise-peer-urls: http://$private_ipv4:2380

**Comment by xiang90** [comment]
@eaoliver 

etcdctl and etcd are totally different things. I think you have to learn more about etcd before using it.

etcdctl is simply a command line tool. You can view it as a client. 

Once you have got a cluster, the best way to config your command line tool is to give it the whole cluster information. 

If you simply use the default, it will only be able to connect the localhost client url. So you have to make sure your localhost etcd is running correctly. Or no matter how well the cluster is running, you WILL NOT be able to connect them, since you never ask it to.

**Comment by eaoliver** [comment]
@xiang90 So the etcd server is broken since we know CoreOS ships with both etcd2 and etcdctl...

**Comment by xiang90** [comment]
@eaoliver 

Let me try again.

1. Whatever CoreOS ships has nothing to do with your problem. Your problem is a runtime issue.

2. If you are using discovery, the cluster WILL NOT be bootstrapped until all the nodes are alive. This is by design, and we will not change it. If you want more flexibility, you can write your own thing via runtime configuration API.

3. After you starting all your members, the cluster will be bootstrapped. At this point, etcdctl list member should work if you give it correct cluster endpoints. If it does not work, then it is a bug. We are aware of a issue around and and are fixing.

4. etcdclt returned you `deadline` error. This is not a good UX and it has a few bugs around. We need to fix it.

5. I suggest you to understand `etcdctl` and `etcd` better before using it. A good start is to read the API documentation. `etcdctl` simply acts as a client, and it needs to know where the cluster is. If you tell it a wrong cluster endpoint, it can never work for you. This is what I suggested you in my 2nd reply.

**Comment by xiang90** [comment]
@eaoliver I am closing this due to low activity. Let me know if you still have the issue or question. Thanks!

**Comment by pjhdld** [comment]
> @eaoliver
> 
> Let me try again.
> 
> 1. Whatever CoreOS ships has nothing to do with your problem. Your problem is a runtime issue.
> 2. If you are using discovery, the cluster WILL NOT be bootstrapped until all the nodes are alive. This is by design, and we will not change it. If you want more flexibility, you can write your own thing via runtime configuration API.
> 3. After you starting all your members, the cluster will be bootstrapped. At this point, etcdctl list member should work if you give it correct cluster endpoints. If it does not work, then it is a bug. We are aware of a related issue and and are fixing it.
> 4. etcdclt returned you `deadline` error. This is not a good UX and it has a few bugs around. We need to fix it.
> 5. I suggest you to understand `etcdctl` and `etcd` better before using it. A good start is to read the API documentation. `etcdctl` simply acts as a client, and it needs to know where the cluster is. If you tell it a wrong cluster endpoint, it can never work for you. This is what I suggested you in my 2nd reply.

@xiang90  I want to know which version of the bug you said is fixed, I seem to have encountered this problem now.
**member list doesn't respond to non-full cluster ("context deadline exceeded"）**

**Comment by hexfusion** [comment]
@pjhdld this issue is very old, the last comment is almost 4 years ago. Please open a new issue with the etcd version you are using and with steps to reproduce.  

> ("context deadline exceeded")

This Is a generic error which basically means the client has timed out waiting for a response from the server.

**Comment by pjhdld** [comment]
> @pjhdld this issue is very old, the last comment is almost 4 years ago. Please open a new issue with the etcd version you are using and with steps to reproduce.
> 
> > ("context deadline exceeded")
> 
> This Is a generic error which basically means the client has timed out waiting for a response from the server.

@hexfusion   ok,thanks
